### PR TITLE
Publish AZ Metrics from Canary Tests.

### PR DESCRIPTION
**Pull request by orsenthil**
**What type of PR is this?**

Canary Tests exercise API Server Connectivity and Inter AZ communication from pods from various AZs.  We want to capture the AZ Metrics from various AZs.

*This is test only metrics and local only to the tests* There is no user impact with this change. This publishes AZ level metrics from the test.  The metrics will use AZ_ID to publish the connectivity Result. There will be two MetricNames  `"APIServerConnectivity"`, and `"InterAZConnectivity"` published in the namespace - `"NetworkingAZConnectivity"`

**Testing done on this change**:

```
Will run 2 of 17 specs
------------------------------
[BeforeSuite]
/Users/senthilx/go/src/github.com/aws/amazon-vpc-cni-k8s/test/integration/cni/pod_networking_suite_test.go:47
  STEP: creating test namespace @ 01/15/24 17:09:22.389
  STEP: getting the node with the node label key kubernetes.io/os and value linux @ 01/15/24 17:09:22.489
  STEP: verifying more than 1 nodes are present for the test @ 01/15/24 17:09:22.642
  STEP: getting the instance type from node label beta.kubernetes.io/instance-type @ 01/15/24 17:09:22.997
  STEP: getting the network interface details from ec2 @ 01/15/24 17:09:22.997
  STEP: describing the VPC to get the VPC CIDRs @ 01/15/24 17:09:23.062
  STEP: setting the environment variables on the ds to map[WARM_ENI_TARGET:0 WARM_IP_TARGET:3] @ 01/15/24 17:09:23.131
  STEP: getting the aws-node daemon set in namespace kube-system @ 01/15/24 17:09:23.131
  STEP: updating the daemon set with new environment variable @ 01/15/24 17:09:23.266
[BeforeSuite] PASSED [12.956 seconds]
------------------------------
SSS
------------------------------
[STATIC_CANARY] API Server Connectivity from AZs While testing API Server Connectivity Should connect to the API Server
/Users/senthilx/go/src/github.com/aws/amazon-vpc-cni-k8s/test/integration/cni/pod_traffic_across_az_test.go:228
  STEP: creating a server DaemonSet on primary node @ 01/15/24 17:09:33.383
  STEP: getting the node with the node label key kubernetes.io/os and value linux @ 01/15/24 17:09:35.442
Testing API Server https://kubernetes.default.svc/api Connectivity from AZ us-west-2c  AZID usw2-az3
API Server https://kubernetes.default.svc/api Connectivity from AZ us-west-2c was successful.
Testing API Server https://kubernetes.default.svc/api Connectivity from AZ us-west-2a  AZID usw2-az2
API Server https://kubernetes.default.svc/api Connectivity from AZ us-west-2a was successful.
Testing API Server https://kubernetes.default.svc/api Connectivity from AZ us-west-2b  AZID usw2-az1
API Server https://kubernetes.default.svc/api Connectivity from AZ us-west-2b was successful.
Testing API Server https://757CBE67CE5FA5DC0BEA97AF921AADAE.gr7.us-west-2.eks.amazonaws.com/api Connectivity from AZ us-west-2c  AZID usw2-az3
API Server https://757CBE67CE5FA5DC0BEA97AF921AADAE.gr7.us-west-2.eks.amazonaws.com/api Connectivity from AZ us-west-2c was successful.
Testing API Server https://757CBE67CE5FA5DC0BEA97AF921AADAE.gr7.us-west-2.eks.amazonaws.com/api Connectivity from AZ us-west-2a  AZID usw2-az2
API Server https://757CBE67CE5FA5DC0BEA97AF921AADAE.gr7.us-west-2.eks.amazonaws.com/api Connectivity from AZ us-west-2a was successful.
Testing API Server https://757CBE67CE5FA5DC0BEA97AF921AADAE.gr7.us-west-2.eks.amazonaws.com/api Connectivity from AZ us-west-2b  AZID usw2-az1
API Server https://757CBE67CE5FA5DC0BEA97AF921AADAE.gr7.us-west-2.eks.amazonaws.com/api Connectivity from AZ us-west-2b was successful.
  STEP: Deleting the Daemonset. @ 01/15/24 17:09:41.801
• [10.475 seconds]
------------------------------
SSSSS
------------------------------
[STATIC_CANARY] test pod networking While testing connectivity across AZ Should allow TCP traffic across AZs.
/Users/senthilx/go/src/github.com/aws/amazon-vpc-cni-k8s/test/integration/cni/pod_traffic_across_az_test.go:137
  STEP: authorizing security group ingress on instance security group @ 01/15/24 17:09:43.858
  STEP: authorizing security group egress on instance security group @ 01/15/24 17:09:44.129
  STEP: creating a server DaemonSet on primary node @ 01/15/24 17:09:44.38
  STEP: getting the node with the node label key kubernetes.io/os and value linux @ 01/15/24 17:09:46.428
  STEP: checking connection on same node, primary to primary @ 01/15/24 17:09:46.966
Testing Connectivity from Pod IP1 192.168.55.37 (us-west-2b, usw2-az1) to Pod IP2 192.168.4.74 (us-west-2a, usw2-az2)
  verifying connectivity from pod netcat-daemonset-ktf8l on node ip-192-168-33-162.us-west-2.compute.internal with IP 192.168.55.37 to pod netcat-daemonset-zjx6z on node ip-192-168-20-49.us-west-2.compute.internal with IP 192.168.4.74
  stdout:  and stderr: Connection to 192.168.4.74 2273 port [tcp/*] succeeded!

Testing Connectivity from Pod IP1 192.168.55.37 (us-west-2b, usw2-az1) to Pod IP2 192.168.95.123 (us-west-2c, usw2-az3)
  verifying connectivity from pod netcat-daemonset-ktf8l on node ip-192-168-33-162.us-west-2.compute.internal with IP 192.168.55.37 to pod netcat-daemonset-2zjqv on node ip-192-168-88-55.us-west-2.compute.internal with IP 192.168.95.123
  stdout:  and stderr: Connection to 192.168.95.123 2273 port [tcp/*] succeeded!

Testing Connectivity from Pod IP1 192.168.95.123 (us-west-2c, usw2-az3) to Pod IP2 192.168.4.74 (us-west-2a, usw2-az2)
  verifying connectivity from pod netcat-daemonset-2zjqv on node ip-192-168-88-55.us-west-2.compute.internal with IP 192.168.95.123 to pod netcat-daemonset-zjx6z on node ip-192-168-20-49.us-west-2.compute.internal with IP 192.168.4.74
  stdout:  and stderr: Connection to 192.168.4.74 2273 port [tcp/*] succeeded!

Testing Connectivity from Pod IP1 192.168.95.123 (us-west-2c, usw2-az3) to Pod IP2 192.168.55.37 (us-west-2b, usw2-az1)
  verifying connectivity from pod netcat-daemonset-2zjqv on node ip-192-168-88-55.us-west-2.compute.internal with IP 192.168.95.123 to pod netcat-daemonset-ktf8l on node ip-192-168-33-162.us-west-2.compute.internal with IP 192.168.55.37
  stdout:  and stderr: Connection to 192.168.55.37 2273 port [tcp/*] succeeded!

Testing Connectivity from Pod IP1 192.168.4.74 (us-west-2a, usw2-az2) to Pod IP2 192.168.55.37 (us-west-2b, usw2-az1)
  verifying connectivity from pod netcat-daemonset-zjx6z on node ip-192-168-20-49.us-west-2.compute.internal with IP 192.168.4.74 to pod netcat-daemonset-ktf8l on node ip-192-168-33-162.us-west-2.compute.internal with IP 192.168.55.37
  stdout:  and stderr: Connection to 192.168.55.37 2273 port [tcp/*] succeeded!

Testing Connectivity from Pod IP1 192.168.4.74 (us-west-2a, usw2-az2) to Pod IP2 192.168.95.123 (us-west-2c, usw2-az3)
  verifying connectivity from pod netcat-daemonset-zjx6z on node ip-192-168-20-49.us-west-2.compute.internal with IP 192.168.4.74 to pod netcat-daemonset-2zjqv on node ip-192-168-88-55.us-west-2.compute.internal with IP 192.168.95.123
  stdout:  and stderr: Connection to 192.168.95.123 2273 port [tcp/*] succeeded!

  STEP: revoking security group ingress on instance security group @ 01/15/24 17:10:01.454
  STEP: revoking security group egress on instance security group @ 01/15/24 17:10:01.74
  STEP: deleting the Daemonset. @ 01/15/24 17:10:02.033
• [18.237 seconds]
------------------------------
SSSSSSS
------------------------------
[AfterSuite]
/Users/senthilx/go/src/github.com/aws/amazon-vpc-cni-k8s/test/integration/cni/pod_networking_suite_test.go:111
  STEP: deleting test namespace @ 01/15/24 17:10:02.096
  STEP: update environment variables map[AWS_VPC_ENI_MTU:9001 AWS_VPC_K8S_CNI_VETHPREFIX:eni], remove map[IP_COOLDOWN_PERIOD:{} WARM_ENI_TARGET:{} WARM_IP_TARGET:{}] @ 01/15/24 17:10:08.261
  STEP: getting the aws-node daemon set in namespace kube-system @ 01/15/24 17:10:08.261
  STEP: updating the daemon set with new environment variable @ 01/15/24 17:10:08.261
[AfterSuite] PASSED [18.547 seconds]
------------------------------

Ran 2 of 17 Specs in 60.216 seconds
SUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 15 Skipped
PASS
```

### CODE REVIEW

**Review by jdn5126 — COMMENTED** [review]
Change overall looks good, but can we control publishing to CloudWatch via a config variable that is disabled by default? These canaries run as part of GitHub Actions runners and other places, so I think we want to limit when we publish to avoid unexpected errors.

**Comment by orsenthil** [comment]
> >  but can we control publishing to CloudWatch via a config variable that is disabled by default? 

Would an environment variable at the test level be good idea? Because this is test only change, which is focused on static_canary test? We can set the environment variable in the runner script.

**Comment by jdn5126** [comment]
> > > but can we control publishing to CloudWatch via a config variable that is disabled by default?
> 
> Would an environment variable at the test level be good idea? Because this is test only change, which is focused on static_canary test? We can set the environment variable in the runner script.

Either sounds fine to me

**Comment by orsenthil** [comment]
@jdn5126 , I thought, it might be good to add a flag `--publish-cw-metrics` to the test framework instead of introducing an envvar. Please take a look.

File: test/framework/options.go [hunk 1]
```diff
@@ -72,6 +73,7 @@ func (options *Options) BindFlags() {
 	flag.StringVar(&options.PublicRouteTableID, "public-route-table-id", "", "Public route table ID (optional, if specified you must specify all of public/private-subnets, public-route-table-id, and availability-zones)")
 	flag.StringVar(&options.NgK8SVersion, "ng-kubernetes-version", "1.25", `Kubernetes version for self-managed node groups (optional, default is "1.25")`)
 	flag.StringVar(&options.TestImageRegistry, "test-image-registry", "617930562442.dkr.ecr.us-west-2.amazonaws.com", `AWS registry where the e2e test images are stored`)
+	flag.BoolVar(&options.PublicCWMetrics, "publish-cw-metrics", false, "Option to publish cloudwatch metrics from the test.")
```
Comment (jdn5126) [inline test/framework/options.go:76]: nit: looks like a typo, `PublicCWMetrics` -> `PublishCWMetrics`

**Review by jdn5126 — APPROVED** [review]

**Pull request merged**